notebooks/implicit_differentiation.ipynb

```diff
@@ -0,0 +1,561 @@
+    "(exp_e_large, exp_f_large), exp_g_large = jax.value_and_grad(explicit_diff,has_aux=True)(param_dict,R_init,displacement,163954)\n",
+    "print('Energy        : ',exp_e_large)\n",
+    "print('Max_grad_force: ',exp_f_large)\n",
+    "print('Gradient of the energy:')\n",
+    "print(exp_g_large['sigma'][0])\n",
+    "print(exp_g_large['alpha'][0])"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "expressed-lightweight",
+   "metadata": {},
+   "source": [
+    "As you can see our computation fails due to running out of memory when trying to compute the gradient. The reason for that is that for reverse mode differentiation (e.g. ```jax.grad```) the memory consumption grows linearly with respect to the number of optimization steps we have to take since reverse mode differentiation needs to store the whole forward pass in order to compute the gradient in the backwards pass."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "mighty-institution",
+   "metadata": {},
+   "source": [
+    "We could lessen the memory requirements by using a technique called [gradient rematerialization/checkpointing](https://github.com/google/jax/blob/f3c4ae3d8918de3a35cec74fdf24231a77ef0e92/jax/_src/api.py#L2806) for a corresponding increase in computation time. While this strategy works, there exists a better solution for our problem called implicit differentiation."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "abroad-charter",
+   "metadata": {},
+   "source": [
+    "# Implicit Differentiation"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "nominated-bunny",
+   "metadata": {},
+   "source": [
+    "Implicit differentiation gets its name from the [implicit function theorem](https://en.wikipedia.org/wiki/Implicit_function_theorem). This theorem roughly states that when we want to differentiate through a root finding procedure, e.g. find $z$ such that $F(a,z) = 0$, it does not matter how we computed the solution. Instead it is possible to directly differentiate through the solution $z^*$ of our root finding problem using the following formular. Here $\\partial_i$ means we differentiate with respect to the $i$'th argument."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "isolated-drove",
+   "metadata": {},
+   "source": [
+    "$$\n",
+    "\\partial z^*(a) = -[\\partial_1 f(a_0,z_0))]^{-1} \\partial_0 f(a_0,z_0))\n",
+    "$$"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "agricultural-rebel",
+   "metadata": {},
+   "source": [
+    "This expression can be efficiently solved. If you are interested in a derivation and how one could implement this in `jax` then I can highly recommend chapter $2$ of the NeurIPS 2020 [Deep Implicit Layers](https://implicit-layers-tutorial.org/) tutorial.\n",
+    "\n",
+    "In order to use implicit differentiation we first need to rewrite our optimization problem as a root finding procedure. This is easily done since the force is $0$ at the energy minimum.\n",
```

Comment:
Maybe add another sentence of detail here; since the force is zero at the minimum, it is a fixed point of gradient descent.